### ISSUE TITLE
Allow to update runner pod annotations on job start

### ISSUE THREAD

**Issue by lovesegfault** [actions-runner-controller/actions-runner-controller#1335]
**Is your feature request related to a problem? Please describe.**

If you're using an autoscaling cluster, it's nice to have runner pods marked as safe to evict, allowing the cluster autoscaler to move them around as needed. If, however, they get evicted while a job is running, you end up with a spurious CI failure, which is unfortunate.

**Describe the solution you'd like**

It should be possible to specify annotations to set when a job starts running. This way, runner pods are marked safe-to-evict while listening, but can remove that annotation when a job is picked up.

**Describe alternatives you've considered**

I considered writing an action that uses the k8s API directly, and then adding it as the first step of all my jobs. This felt really cumbersome, and added boilerplate to the entire workflow.

**Comment by mumoshu** [comment]
@lovesegfault Hey. Could you try https://github.com/actions-runner-controller/actions-runner-controller/blob/85dea9b67c9ed5d6c634f338cbf28070dd5c84f0/README.md#additional-tweaks? Any runner pod annotations should be propagated down to runner pods. You can customize runner pod annotations via runnerdeployment.spec.template.metadata.annotations in case you use RunnerDeployment. RunnerSet supports similar configurations.

**Comment by lovesegfault** [comment]
@mumoshu That works fine to have an annotation _always_ set for a Deployment, the issue here is changing the value of an annotation when the runner picks up a task.

Essentially, I want to spawn RunnerDeployments with `cluster-autoscaler.kubernetes.io/safe-to-evict: "true"`, but have them change that to `false` when a job starts running.

Without this, you can get evicted halfway through a running job, causing it to fail.

**Comment by mumoshu** [comment]
I've updated the title to more accuretely capture the goal

**Comment by lovesegfault** [comment]
> How is that possible? How do you detect a runner to have started running a job?

Apparently, you can add arbitrary code before a job starts: https://github.com/actions/runner/issues/699

**Comment by mumoshu** [comment]
Or do you imagine anything more sophisticated, that ARC implements another HTTP server that is used to receive our custom HTTP notifications from every runner's job start hook, so that ARC can update annotations of the notification sender?

**Comment by lovesegfault** [comment]
@mumoshu 
> Are you okay with giving every runner pod permission to update the annotation of itself...?

I think that should be fine? I imagine many people using ARC are on private repos where this won't really be a problem.

> Or do you imagine anything more sophisticated, that ARC implements another HTTP server that is used to receive our custom HTTP notifications from every runner's job start hook, so that ARC can update annotations of the notification sender?

That would be pretty cool, especially because it would also let the controller know a runner is now busy without having to hit the GitHub API at all.

**Comment by lovesegfault** [comment]
Oh, I just realized, I think we can have webhook-like autoscaling without webhooks by using these runner hooks!

**Comment by mumoshu** [comment]
> Oh, I just realized, I think we can have webhook-like autoscaling without webhooks by using these runner hooks!

How?! You need a runner pod and a job to trigger the notification in the first place. That basically means the speed of scale up is limited by the current number of runner pods(that are capable on sending notifications) and assumption that you're ok with adding runner pods before any jobs are queued.
Our webhook autoscaler reacts on `queued` workflow_job events so in a good scenario it adds enough and sufficient runners, whereas anything based on job start hook would work like the opposite manner(=adds more runner pods as existing runner pods gets busy).

**Comment by mumoshu** [comment]
OK I just saw your updated comment :)

**Comment by lovesegfault** [comment]
> Yeah... I tend to think so but at the same time some folks have started asking to remove privileged from runner pods, which shouldn't be an issue if you're mainly running ARC in a controlled environment(including your point, use it only on private repos).

Right, it's much easier to do things if you don't have to worry about someone attacking the runners.

Regarding `privileged`, I have to force-enable it for us; otherwise the Nix installer doesn't work. You can repro it by ssh'ing into a runner and running:

```console
$ sh <(curl -L https://nixos.org/nix/install) --no-daemon
```

I'm sure others have similar issues where they need tooling that interacts negatively with heavy-handed things like SELinux.

Anyway, my point is just that the simple solution is one that only suits private repos.

> > That would be pretty cool, especially because it would also let the controller know a runner is now busy without having to hit the GitHub API at all.
> 
> Maybe! I'm still not sure how reliable it is though we may try implement it behind a feature flag if we go that way.

There are other fancy things that this enables, like letting the cluster know how long a certain job took.

If ARC keeps track of that long-term, it can use it to influence autoscaling decisions, which is powerful.

I imagine it would be hard to implement such a system, however.

**Comment by mumoshu** [comment]
> I want to spawn RunnerDeployments with cluster-autoscaler.kubernetes.io/safe-to-evict: "true"

This turns out to be unsafe, because there seem to no way to prevent ARC and cluster-autoscaler from racing, like in a scenario that CA starts the node termination process seeing `safe-to-evict: "true` but at the same time `actions/runner` in a runner pod starts running a job(because it doesn't know that the runner pod's node is being terminated).

**Comment by mumoshu** [comment]
ARC implements a graceful scale-down logic that basically works by unregistering the runner first and then mark the runner pod for deletion. This way, a runner pod is guaranteed to not running any job when it was requested to terminate.

Do you know any cluster-autoscaler feature that can be used to notify ARC that CA wants to scale down a node/pod? If there's one, ARC could unregister the runner first and sets `safe-to-evict: "true"` only after that, then CA start the node termination process.

**Comment by lovesegfault** [comment]
> Do you know any cluster-autoscaler feature that can be used to notify ARC that CA wants to scale down a node/pod? If there's one, ARC could unregister the runner first and sets `safe-to-evict: "true"` only after that, then CA start the node termination process.

I don't know how it works, but it must exist because GKE can tell me which nodes are blocking scale-down. It will even tell me marking `safe-to-evict` might help.

**Comment by lovesegfault** [comment]
Looking through this: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md

**Comment by mumoshu** [comment]
@lovesegfault Thanks. Yeah, I read it a few times and my understanding is that it's blocked whenever a pod that is not evict-safe is there. But I think ARC's use-case is a bit different because to avoid race we need to detect if CA wants to scale down, and only after that unblock the node for termination.

**Comment by lovesegfault** [comment]
> to avoid race we need to detect if CA wants to scale down, and only after that unblock the node for termination.

Looking here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-events-are-emitted-by-ca

> on pods:
>    - TriggeredScaleUp - CA decided to scale up cluster to make place for this pod.
>    - NotTriggerScaleUp - CA couldn't find node group that can be scaled up to make this pod schedulable.
>    - ScaleDown - CA will try to evict this pod as part of draining the node.

I wonder whether that `ScaleDown` event is enough, the controller can then initiate graceful termination when it sees that.

According to this: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#does-ca-respect-gracefultermination-in-scale-down

That should work, no?

**Comment by lovesegfault** [comment]
So:

1. CA emits a ScaleDown event on the Pod
2. ARC sees that and:
  1. If the pod is running, does nothing
  2. If the pod is not running, starts shutting it down
3. The pod terminates gracefully
4. CA scales down the node

I imagine CA cordons off the node before issuing `ScaleDown` on pods, so it won't just get re-scheduled to the same node.

**Comment by mumoshu** [comment]
> According to this: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#does-ca-respect-gracefultermination-in-scale-down

It's close but doesn't perfectly fit ARC's use-case. The graceful termination period needs to be longer than what's allowed by pod spec's terminationGracePeriodSeconds, so as CA's max grace period, because a workflow job can be running longer than that. That's why ARC uses its own graceful termination logic outside of the K8s finalizer(which can't block pod containers from existing when terminationGracePeriodSeconds exceeded).

**Comment by mumoshu** [comment]
> I wonder whether that ScaleDown event is enough, the controller can then initiate graceful termination when it sees that.

That's not enough, because CA creates the event only after it decided to scale down, and at that time CA should have already started the grace period, which isn't what we want here.

**Comment by lovesegfault** [comment]
Can we still get CA events if the pods _aren't_ marked safe-to-evict? Then ARC can be the one to handle termination iff the pod is not running a job.

IOW: Can the question "Is this pod blocking a CA scale-down" be answered from ARC's POV?

**Comment by mumoshu** [comment]
> Can we still get CA events if the pods aren't marked safe-to-evict? Then ARC can be the one to handle termination iff the pod is not running a job.

What do you mean by this?
If it wasn't marked as safe-to-evict, CA would think the node is unnecessary and therefore start the node termination process, which involves draining of pods, and the pod drains are subject to terminationGracePeriodSeconds and node termination is subject to CA's max grace period.

> IOW: Can the question "Is this pod blocking a CA scale-down" be answered from ARC's POV?

My answer is "no", and that's why I asked https://github.com/actions-runner-controller/actions-runner-controller/issues/1335#issuecomment-1098625597.

**Comment by mumoshu** [comment]
I'd suggest reading this, if you're curious
https://github.com/kubernetes/autoscaler/blob/a9ec278ad121b684205661d0270acae86c8da777/cluster-autoscaler/core/scale_down.go#L787-L789

From ARC's perspective, we'd want to let CA to mark a node or a pod for potential termination (with no timeout) first, so that ARC can remark those pods as safe-to-evict after un-registrations. From what I can see from CA's code, I don't see any way to achieve that...

**Comment by mumoshu** [comment]
If it isn't possible in cooperation with CA, another approach would be to implement our own variant of [descheduler](https://github.com/kubernetes-sigs/descheduler) and tweak K8s scheduler to bin-pack runner pods. If the variant of descheduler proactively descheduled runner pods on low-utilized nodes, the K8s scheduler would eventually schedule runner pods to only a subset of nodes, and therefore CA detects now-empty nodes for scale-down.

**Comment by lovesegfault** [comment]
I'm going to try to loop-in some of the K8S people from work in here to see if we're missing anything :)

**Comment by AndreLacasse** [comment]
Is there any update about this issue because de only solution I find is to have a minimum of 0 runner.

**Comment by mumoshu** [comment]
@AndreLacasse Since then the Actions team has added a feature called job hooks to actions runner. Check out https://docs.github.com/ja/actions/hosting-your-own-runners/running-scripts-before-or-after-a-job. I think it might be useful for your use-case.
